chat-widget/src/components/draggable/DraggableChatWidget.tsx

```diff
@@ -0,0 +1,166 @@
+    const [cachedPosition, setCachedPosition] = useState<undefined|DraggableElementPosition>(undefined);
+    const [position, setPosition] = useState({offsetLeft: 0, offsetTop: 0});
+    const [delta, setDelta] = useState({left: 0, top: 0});
+
+    const repositionElement = (draggableElement: HTMLElement, offsetLeft: number, offsetTop: number) => {
+        draggableElement.style.left = `${offsetLeft}px`;
+        draggableElement.style.top = `${offsetTop}px`;
+    };
+
+    const calculateOffsetsWithinViewport = useCallback((id: string, offset, delta) => {
+        const draggableElement: HTMLElement | null = document.getElementById(id);
+        const positionRelativeToViewport = (draggableElement as HTMLElement).getBoundingClientRect();
+
+        let offsetLeft = offset.offsetLeft;
+        let offsetTop = offset.offsetTop;
+
+        // Widget size larger than viewport would not have any restriction
+        if (positionRelativeToViewport.width > window.innerWidth) {
+            return;
+        }
+
+        if (positionRelativeToViewport.height > window.innerHeight) {
+            return;
+        }
+
+        // Ensures widget is within viewport
+        if (positionRelativeToViewport.x < 0) {
+            offsetLeft = 0 - delta.left;
+        }
+
+        if (positionRelativeToViewport.y < 0) {
+            offsetTop = 0 - delta.top;
+        }
+
+        if (positionRelativeToViewport.x + positionRelativeToViewport.width > window.innerWidth) {
+            offsetLeft = window.innerWidth - positionRelativeToViewport.width - delta.left;
+        }
+
+        if (positionRelativeToViewport.y + positionRelativeToViewport.height > window.innerHeight) {
+            offsetTop = window.innerHeight - positionRelativeToViewport.height - delta.top;
+        }
+
+        repositionElement(draggableElement as HTMLElement, offsetLeft, offsetTop);
+        setPosition({offsetLeft, offsetTop});
+    }, []);
+
+    const resetPosition = useCallback((targetPosition: DraggableElementPosition) => {
+        calculateOffsetsWithinViewport(props.elementId, targetPosition, delta); // Ensure viewport restriction
+    }, [delta]);
+
+    useEffect(() => {
+        if (props.disable === true) {
+            return;
+        }
+
+        const cacheInitialPosition = () => {
+            const draggableElement: HTMLElement | null = document.getElementById(props.elementId);
+            const offsetLeft = (draggableElement as HTMLElement).offsetLeft as number;
+            const offsetTop = (draggableElement as HTMLElement).offsetTop as number;
+            setInitialPosition({offsetLeft, offsetTop});
+        };
+
+        const calculateOffsets = () => {
+            const draggableElement: HTMLElement | null = document.getElementById(props.elementId);
+            const offsetLeft = (draggableElement as HTMLElement).offsetLeft as number;
+            const offsetTop = (draggableElement as HTMLElement).offsetTop as number;

```

Comment:
This seems to be duplicate code. Can we move it out above line 77?

Reply:
It's just retrieving the offsets of the `draggableElement`. It should be fine